### ISSUE TITLE
Memory leak in generic_resize/reallocate_copy

### ISSUE THREAD

**Issue by nickmertin** [dimforge/nalgebra#1378]
I believe there is a memory leak arising from `DefaultAllocator::reallocate_copy` and its use in `Matrix::resize_generic`. Anecdotally, I observed memory leakage in the wild in some pre-production code and narrowed its source down to here.

The resulting matrix is created by this line of `resize_generic`:

```rust
let res = unsafe { DefaultAllocator::reallocate_copy(new_nrows, new_ncols, data.data) };
```

In the case where at least one of the dimensions of the input matrix to `resize_generic` is `Dyn`, it is backed (by default) by a `VecStorage`; this is the type of `data.data` in the above line. However, in the case where the output matrix is statically sized, `reallocate_copy` moves the data out of this buffer to the new one, and then forgets the buffer:

```rust
// Safety:
// - We don’t care about dropping elements because the caller is responsible for dropping things.
// - We forget `buf` so that we don’t drop the other elements.
std::mem::forget(buf);
```

I believe the intention here is to ensure that the destructor does not get called on the *elements* of the old matrix, as the ones being used are moved to the new buffer, and the others are explicitly dropped by `resize_generic`; the latter part is, I believe, what is referred to by "the caller is responsible for dropping things." However, the caller cannot be responsible for dropping the buffer itself, as it is consumed by `reallocate_copy`; that is the source of the issue here.

I will try to put together a PR shortly to fix this and ensure that the allocation gets cleaned up, while maintaining memory safety regarding not double-dropping the matrix elements.

**Comment by mert-kurttutan** [comment]
I was just reviewing, some issues.
 Shouldn't this be closed by #1416 ?

**Comment by nickmertin** [comment]
Yes, closed by #1416.
